apps/hash-ai-worker-ts/src/activities/flow-activities/shared/get-file-entity-from-url.ts

```diff
@@ -0,0 +1,273 @@
+    );
+  }
+};
+
+const writeFileToS3URL = async ({
+  filePath,
+  mimeType,
+  presignedPutUrl,
+  sizeInBytes,
+}: {
+  filePath: string;
+  mimeType: string;
+  presignedPutUrl: string;
+  sizeInBytes: number;
+}) => {
+  return new Promise((resolve, reject) => {
+    const fileStream = createReadStream(filePath);
+    const req = (presignedPutUrl.startsWith("https://") ? https : http).request(
+      presignedPutUrl,
+      {
+        headers: { "Content-Length": sizeInBytes, "Content-Type": mimeType },
+        method: "PUT",
+      },
+      (res) => {
+        res.on("data", () => {});
+        res.on("end", () => {
+          if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
+            resolve("Ok");
+          } else {
+            reject(
+              new Error(
+                `${res.statusCode} Error uploading to S3: ${res.statusMessage}`,
+              ),
+            );
+          }
+        });
+      },
+    );
+
+    req.on("error", (error) => {
+      reject(new Error(`Request error: ${error.message}`));
+    });
+
+    fileStream.on("error", (error) =>
+      reject(new Error(`Error reading file: ${error.message}`)),
+    );
+    fileStream.pipe(req);
+    fileStream.on("end", () => req.end());
+  }).finally(() => {
+    void unlink(filePath);
+  });
+};
+
+export const getFileEntityFromUrl = async (params: {
+  url: string;
+  entityTypeId?: VersionedUrl;
+  description?: string;
+  displayName?: string;
+}): Promise<
+  | {
+      status: "ok";
+      entityMetadata: EntityMetadata;
+      properties: FileProperties;
+    }
+  | {
+      status: "error-uploading-file";
+      message: string;
+    }
+  | {
+      status: "error-downloading-file";
+      message: string;
+    }
+> => {
+  const { url: originalUrl, description, displayName } = params;
+
+  const { userAuthentication } = await getFlowContext();
+
+  const urlObject = new URL(originalUrl);
+  const urlWithoutParams = new URL(urlObject.origin + urlObject.pathname);
+  const filename = urlWithoutParams.pathname.split("/").pop()!;
+
+  let localFilePath;
+  try {
+    localFilePath = await downloadFileToFileSystem(originalUrl);
+  } catch (err) {
+    const message = `Error downloading file from URL: ${(err as Error).message}`;
+
+    // @todo logger – update once #4290 merges
+    // eslint-disable-next-line no-console
+    console.log(message);
+    return {
+      status: "error-downloading-file",
+      message,
+    };
+  }
+
+  const mimeType = mime.lookup(filename) || "application/octet-stream";
+  const entityTypeId =
+    params.entityTypeId ??
+    getEntityTypeIdForMimeType(mimeType) ??
+    systemEntityTypes.file.entityTypeId;
+
+  const stats = statSync(localFilePath);
+  const fileSizeInBytes = stats.size;
+
+  /** @todo: allow overriding this via an input, as with other actions (hardcoded in Flow definition?) */
+  const ownedById = userAuthentication.actorId;
```

Comment:
This can be taken from context now

Reply:
Addressed in https://github.com/hashintel/hash/pull/4399/commits/06508102ebde6e372441fc4b12d3c00373352c6e